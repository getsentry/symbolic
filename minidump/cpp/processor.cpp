#include "google_breakpad/processor/basic_source_line_resolver.h"
#include "google_breakpad/processor/minidump.h"
#include "google_breakpad/processor/minidump_processor.h"
#include "google_breakpad/processor/process_state.h"

#include "cpp/data_definitions.h"
#include "cpp/memstream.h"
#include "cpp/mmap_symbol_supplier.h"
#include "cpp/processor.h"

using google_breakpad::BasicSourceLineResolver;
using google_breakpad::Minidump;
using google_breakpad::MinidumpProcessor;
using google_breakpad::ProcessState;

process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out) {
    if (buffer == nullptr) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    ProcessState *state = new ProcessState();
    if (state == nullptr) {
        *result_out = -1;  // Memory allocation issue
        return nullptr;
    }

    BasicSourceLineResolver resolver;
    MmapSymbolSupplier supplier(symbol_count, symbols);
    MinidumpProcessor processor(&supplier, &resolver);

    imemstream in(buffer, buffer_size);
    Minidump minidump(in);
    if (!minidump.Read()) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    *result_out = processor.Process(&minidump, state);
    if (*result_out != google_breakpad::PROCESS_OK) {
        delete state;
        return nullptr;
    }

    return process_state_t::cast(state);
}
