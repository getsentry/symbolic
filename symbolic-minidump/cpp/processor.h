#ifndef SENTRY_PROCESSOR_H
#define SENTRY_PROCESSOR_H

#include <cstddef>
#include "cpp/data_structures.h"

#ifdef __cplusplus
extern "C" {
#endif

/// Data transfer object for symbols in memory
struct symbol_entry_t {
    /// The debug identifier of the code module these symbols are for
    const char *debug_identifier;

    /// Size of the buffer inside symbol_data
    const size_t symbol_size;

    /// Raw data of the symbol file passed to the symbolizer
    const char *symbol_data;
};

/// Reads a minidump from a memory buffer and processes it. Returns an owning
/// pointer to a process_state_t struct that contains loaded code modules and
/// call stacks of all threads of the process during the crash.
///
/// Processing the minidump can fail if the buffer is corrupted or does not
/// exit. The function will return NULL and an error code in result_out.
///
/// Release memory of the process state with process_state_delete.
process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out);

#ifdef __cplusplus
}
#endif

#endif
