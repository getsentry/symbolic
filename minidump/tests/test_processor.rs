use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_minidump::processor::ProcessState;

#[test]
fn process_minidump_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_linux", &state);
    Ok(())
}

#[test]
fn process_minidump_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_macos", &state);
    Ok(())
}

#[test]
fn process_minidump_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_windows", &state);
    Ok(())
}

#[test]
fn get_referenced_modules_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_linux", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_macos", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!(
        "referenced_modules_windows",
        &state.referenced_modules()
    );
    Ok(())
}
