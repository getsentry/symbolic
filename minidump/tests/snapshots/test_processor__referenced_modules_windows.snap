---
created: "2019-02-20T15:42:12.012672Z"
creator: insta@0.6.3
source: minidump/tests/test_processor.rs
expression: "&state.referenced_modules()"
---
{
    CodeModule {
        id: Some(
            CodeModuleId {
                inner: DebugId {
                    uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
                    appendix: 1
                }
            }
        ),
        base_address: 2752512,
        size: 36864,
        code_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe",
        code_identifier: "5AB380779000",
        debug_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb",
        debug_identifier: "3249D99D0C4049318610F4E4FB0B69361"
    },
    CodeModule {
        id: Some(
            CodeModuleId {
                inner: DebugId {
                    uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                    appendix: 1
                }
            }
        ),
        base_address: 1997996032,
        size: 1585152,
        code_file: "C:\\Windows\\System32\\ntdll.dll",
        code_identifier: "59B0D8F3183000",
        debug_file: "wntdll.pdb",
        debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
    },
    CodeModule {
        id: Some(
            CodeModuleId {
                inner: DebugId {
                    uuid: "ae131c67-27a7-4fa1-9916-b5a4aef41190",
                    appendix: 1
                }
            }
        ),
        base_address: 1971388416,
        size: 790528,
        code_file: "C:\\Windows\\System32\\rpcrt4.dll",
        code_identifier: "5A49BB75c1000",
        debug_file: "wrpcrt4.pdb",
        debug_identifier: "AE131C6727A74FA19916B5A4AEF411901"
    },
    CodeModule {
        id: Some(
            CodeModuleId {
                inner: DebugId {
                    uuid: "aec7ef2f-df4b-4642-a471-4c3e5fe8760a",
                    appendix: 1
                }
            }
        ),
        base_address: 1891041280,
        size: 151552,
        code_file: "C:\\Windows\\System32\\dbgcore.dll",
        code_identifier: "57898DAB25000",
        debug_file: "dbgcore.pdb",
        debug_identifier: "AEC7EF2FDF4B4642A4714C3E5FE8760A1"
    },
    CodeModule {
        id: Some(
            CodeModuleId {
                inner: DebugId {
                    uuid: "d3474559-96f7-47d6-bf43-c176b2171e68",
                    appendix: 1
                }
            }
        ),
        base_address: 1963261952,
        size: 917504,
        code_file: "C:\\Windows\\System32\\kernel32.dll",
        code_identifier: "590285E9e0000",
        debug_file: "wkernel32.pdb",
        debug_identifier: "D347455996F747D6BF43C176B2171E681"
    }
}
